File: bindings/typescript/tests/validation.test.ts
/**
 * Tests for client-side ContentPath, MimeType, and deploy_id validation (SCP-297).
 *
 * Validates that the SDK-layer validation functions produce clear, descriptive
 * error messages for invalid inputs BEFORE the FFI boundary is crossed.
 *
 * Mirrors the Rust validation in `crates/scp-core/src/context/broadcast_content.rs`.
 */

import { describe, expect, it } from "bun:test";
import { _validateContentPath, _validateDeployId, _validateMimeType } from "../src/context";
import { ValidationError } from "../src/errors";

// ---------------------------------------------------------------------------
// ContentPath validation
// ---------------------------------------------------------------------------

describe("validateContentPath (SCP-297)", () => {
  it("accepts root path", () => {
    expect(() => _validateContentPath("/")).not.toThrow();
  });

  it("accepts simple path", () => {
    expect(() => _validateContentPath("/index.html")).not.toThrow();
  });

  it("accepts nested path", () => {
    expect(() => _validateContentPath("/assets/css/main.css")).not.toThrow();
  });

  it("accepts hidden file", () => {
    expect(() => _validateContentPath("/.well-known/acme-challenge/token")).not.toThrow();
  });

  it("rejects path without leading slash", () => {
    expect(() => _validateContentPath("index.html")).toThrow(ValidationError);
    expect(() => _validateContentPath("index.html")).toThrow(/must start with '\/'/);
  });

  it("rejects empty string", () => {
    expect(() => _validateContentPath("")).toThrow(ValidationError);
  });

  it("rejects path exceeding 1024 bytes", () => {
    expect(() => _validateContentPath(`/${"a".repeat(1024)}`)).toThrow(/exceeds 1024 bytes/);
  });

  it("rejects backslash", () => {
    expect(() => _validateContentPath("/path\\file")).toThrow(/backslashes/);
  });

  it("rejects percent-encoded bytes", () => {
    expect(() => _validateContentPath("/path%20file")).toThrow(/percent-encoded/);
  });

  it("rejects query string", () => {
    expect(() => _validateContentPath("/path?key=value")).toThrow(/query strings/);
  });

  it("rejects fragment", () => {
    expect(() => _validateContentPath("/path#section")).toThrow(/fragments/);
  });

  it("rejects null byte", () => {
    expect(() => _validateContentPath("/path\0file")).toThrow(/null bytes/);
  });

  it("rejects control character", () => {
    expect(() => _validateContentPath("/path\x01file")).toThrow(/control character/);
  });

  it("rejects DEL character", () => {
    expect(() => _validateContentPath("/path\x7ffile")).toThrow(/control character/);
  });

  it("rejects double slash", () => {
    expect(() => _validateContentPath("/path//file")).toThrow(/\/\//);
  });

  it("rejects trailing slash", () => {
    expect(() => _validateContentPath("/path/")).toThrow(/trailing slash/);
  });

  it("rejects dot segment", () => {
    expect(() => _validateContentPath("/path/./file")).toThrow(/'\.' segments/);
  });

  it("rejects dotdot segment", () => {
    expect(() => _validateContentPath("/path/../etc/passwd")).toThrow(/directory traversal/);
  });

  it("rejects C1 control character (Fix 4)", () => {
    expect(() => _validateContentPath("/path\u0085file")).toThrow(/control character U\+0085/);
  });

  it("rejects zero-width space (Fix 1)", () => {
    expect(() => _validateContentPath("/path\u200Bfile")).toThrow(/whitespace\/formatting U\+200B/);
  });

  it("rejects bidi override (Fix 1)", () => {
    expect(() => _validateContentPath("/path\u202Efile")).toThrow(/whitespace\/formatting U\+202E/);
  });

  it("rejects NBSP (Fix 1)", () => {
    expect(() => _validateContentPath("/path\u00A0file")).toThrow(/whitespace\/formatting U\+00A0/);
  });

  it("NFC normalizes before validation (Fix 3)", () => {
    // U+0065 U+0301 (e + combining acute) normalizes to U+00E9 (e-acute)
    expect(() => _validateContentPath("/caf\u0065\u0301")).not.toThrow();
  });

  it("uses error code SCP-VALID-7010", () => {
    expect.assertions(2);
    try {
      _validateContentPath("no-slash");
    } catch (e) {
      expect(e).toBeInstanceOf(ValidationError);
      expect((e as ValidationError).code).toBe("SCP-VALID-7010");
    }
  });
});

// ---------------------------------------------------------------------------
// MimeType validation
// ---------------------------------------------------------------------------

describe("validateMimeType (SCP-297)", () => {
  it("accepts text/html", () => {
    expect(() => _validateMimeType("text/html")).not.toThrow();
  });

  it("accepts application/json", () => {
    expect(() => _validateMimeType("application/json")).not.toThrow();
  });

  it("accepts image/png", () => {
    expect(() => _validateMimeType("image/png")).not.toThrow();
  });

  it("rejects empty string", () => {
    expect(() => _validateMimeType("")).toThrow(/must not be empty/);
  });

  it("rejects no slash", () => {
    expect(() => _validateMimeType("texthtml")).toThrow(/exactly one '\/'/);
  });

  it("rejects double slash", () => {
    expect(() => _validateMimeType("text/html/extra")).toThrow(/exactly one '\/'/);
  });

  it("rejects empty type part", () => {
    expect(() => _validateMimeType("/html")).toThrow(/both be non-empty/);
  });

  it("rejects empty subtype part", () => {
    expect(() => _validateMimeType("text/")).toThrow(/both be non-empty/);
  });

  it("rejects semicolon (parameters)", () => {
    expect(() => _validateMimeType("text/html; charset=utf-8")).toThrow(/parameters/);
  });

  it("rejects carriage return", () => {
    expect(() => _validateMimeType("text/html\r")).toThrow(/control character/);
  });

  it("rejects line feed", () => {
    expect(() => _validateMimeType("text/html\n")).toThrow(/control character/);
  });

  it("rejects null byte", () => {
    expect(() => _validateMimeType("text/\x00html")).toThrow(/control character/);
  });

  it("rejects C1 control character in MIME type (Fix 4)", () => {
    expect(() => _validateMimeType("text/\u0085html")).toThrow(/control character U\+0085/);
  });

  it("rejects non-tchar in type part (Fix 2)", () => {
    expect(() => _validateMimeType("te xt/html")).toThrow(/type part contains invalid/);
  });

  it("rejects non-tchar in subtype part (Fix 2)", () => {
    expect(() => _validateMimeType("text/ht ml")).toThrow(/subtype part contains invalid/);
  });

  it("accepts tchar special characters (Fix 2)", () => {
    expect(() => _validateMimeType("application/vnd.foo+bar")).not.toThrow();
  });

  it("uses error code SCP-VALID-7011", () => {
    expect.assertions(2);
    try {
      _validateMimeType("");
    } catch (e) {
      expect(e).toBeInstanceOf(ValidationError);
      expect((e as ValidationError).code).toBe("SCP-VALID-7011");
    }
  });
});

// ---------------------------------------------------------------------------
// deploy_id validation
// ---------------------------------------------------------------------------

describe("validateDeployId (SCP-297)", () => {
  it("accepts simple deploy ID", () => {
    expect(() => _validateDeployId("deploy-1")).not.toThrow();
  });

  it("accepts hex string", () => {
    expect(() => _validateDeployId("abc123def456")).not.toThrow();
  });

  it("accepts underscore", () => {
    expect(() => _validateDeployId("my_deploy_id")).not.toThrow();
  });

  it("accepts mixed case", () => {
    expect(() => _validateDeployId("Deploy-2024_v1")).not.toThrow();
  });

  it("rejects empty string", () => {
    expect(() => _validateDeployId("")).toThrow(/must not be empty/);
  });

  it("rejects string over 128 bytes", () => {
    expect(() => _validateDeployId("a".repeat(129))).toThrow(/exceeds 128 bytes/);
  });

  it("rejects spaces", () => {
    expect(() => _validateDeployId("deploy 1")).toThrow(/ASCII alphanumeric/);
  });

  it("rejects special characters", () => {
    expect(() => _validateDeployId("deploy@1")).toThrow(/ASCII alphanumeric/);
  });

  it("rejects slash", () => {
    expect(() => _validateDeployId("deploy/1")).toThrow(/ASCII alphanumeric/);
  });

  it("uses error code SCP-VALID-7012", () => {
    expect.assertions(2);
    try {
      _validateDeployId("");
    } catch (e) {
      expect(e).toBeInstanceOf(ValidationError);
      expect((e as ValidationError).code).toBe("SCP-VALID-7012");
    }
  });
});

File: bindings/kotlin/scp-kt/src/test/kotlin/works/limn/scp/ValidationTest.kt
// Tests for client-side ContentPath, MimeType, and deploy_id validation (SCP-297).
//
// Validates that the SDK-layer validation functions produce clear, descriptive
// error messages for invalid inputs BEFORE the FFI boundary is crossed.
//
// Mirrors the Rust validation in `crates/scp-core/src/context/broadcast_content.rs`.
//
// Provenance: SCP-297, spec §18.11.9

package works.limn.scp

import org.junit.jupiter.api.Test
import org.junit.jupiter.api.assertDoesNotThrow
import org.junit.jupiter.api.assertThrows
import works.limn.scp.bridge.BridgeException
import kotlin.test.assertContains
import kotlin.test.assertEquals

class ValidationTest {
    // -- ContentPath ----------------------------------------------------------

    @Test
    fun `content path - valid root path accepted`() {
        assertDoesNotThrow { validateContentPath("/") }
    }

    @Test
    fun `content path - valid simple path accepted`() {
        assertDoesNotThrow { validateContentPath("/index.html") }
    }

    @Test
    fun `content path - valid nested path accepted`() {
        assertDoesNotThrow { validateContentPath("/assets/css/main.css") }
    }

    @Test
    fun `content path - valid hidden file accepted`() {
        assertDoesNotThrow { validateContentPath("/.well-known/acme-challenge/token") }
    }

    @Test
    fun `content path - rejects no leading slash`() {
        val ex = assertThrows<BridgeException> { validateContentPath("index.html") }
        assertContains(ex.message.orEmpty(), "must start with '/'")
        assertEquals("SCP-VALID-7010", ex.code)
    }

    @Test
    fun `content path - rejects empty string`() {
        assertThrows<BridgeException> { validateContentPath("") }
    }

    @Test
    fun `content path - rejects too long`() {
        val ex = assertThrows<BridgeException> { validateContentPath("/" + "a".repeat(1024)) }
        assertContains(ex.message.orEmpty(), "exceeds 1024 bytes")
    }

    @Test
    fun `content path - rejects backslash`() {
        val ex = assertThrows<BridgeException> { validateContentPath("/path\\file") }
        assertContains(ex.message.orEmpty(), "backslashes")
    }

    @Test
    fun `content path - rejects percent-encoded bytes`() {
        val ex = assertThrows<BridgeException> { validateContentPath("/path%20file") }
        assertContains(ex.message.orEmpty(), "percent-encoded")
    }

    @Test
    fun `content path - rejects query string`() {
        val ex = assertThrows<BridgeException> { validateContentPath("/path?key=value") }
        assertContains(ex.message.orEmpty(), "query strings")
    }

    @Test
    fun `content path - rejects fragment`() {
        val ex = assertThrows<BridgeException> { validateContentPath("/path#section") }
        assertContains(ex.message.orEmpty(), "fragments")
    }

    @Test
    fun `content path - rejects null byte`() {
        val ex = assertThrows<BridgeException> { validateContentPath("/path\u0000file") }
        assertContains(ex.message.orEmpty(), "null bytes")
    }

    @Test
    fun `content path - rejects control character`() {
        val ex = assertThrows<BridgeException> { validateContentPath("/path\u0001file") }
        assertContains(ex.message.orEmpty(), "control character")
    }

    @Test
    fun `content path - rejects DEL character`() {
        val ex = assertThrows<BridgeException> { validateContentPath("/path\u007Ffile") }
        assertContains(ex.message.orEmpty(), "control character")
    }

    @Test
    fun `content path - rejects double slash`() {
        val ex = assertThrows<BridgeException> { validateContentPath("/path//file") }
        assertContains(ex.message.orEmpty(), "'//'")
    }

    @Test
    fun `content path - rejects trailing slash`() {
        val ex = assertThrows<BridgeException> { validateContentPath("/path/") }
        assertContains(ex.message.orEmpty(), "trailing slash")
    }

    @Test
    fun `content path - rejects dot segment`() {
        val ex = assertThrows<BridgeException> { validateContentPath("/path/./file") }
        assertContains(ex.message.orEmpty(), "'.' segments")
    }

    @Test
    fun `content path - rejects dotdot segment`() {
        val ex = assertThrows<BridgeException> { validateContentPath("/path/../etc/passwd") }
        assertContains(ex.message.orEmpty(), "directory traversal")
    }

    @Test
    fun `content path - rejects C1 control character`() {
        val ex = assertThrows<BridgeException> { validateContentPath("/path\u0085file") }
        assertContains(ex.message.orEmpty(), "control character U+0085")
    }

    @Test
    fun `content path - rejects zero-width space`() {
        val ex = assertThrows<BridgeException> { validateContentPath("/path\u200Bfile") }
        assertContains(ex.message.orEmpty(), "whitespace/formatting U+200B")
    }

    @Test
    fun `content path - rejects bidi override`() {
        val ex = assertThrows<BridgeException> { validateContentPath("/path\u202Efile") }
        assertContains(ex.message.orEmpty(), "whitespace/formatting U+202E")
    }

    @Test
    fun `content path - rejects NBSP`() {
        val ex = assertThrows<BridgeException> { validateContentPath("/path\u00A0file") }
        assertContains(ex.message.orEmpty(), "whitespace/formatting U+00A0")
    }

    @Test
    fun `content path - NFC normalizes before validation`() {
        // U+0065 U+0301 (e + combining acute) normalizes to U+00E9 (e-acute)
        assertDoesNotThrow { validateContentPath("/caf\u0065\u0301") }
    }

    // -- MimeType -------------------------------------------------------------

    @Test
    fun `mime type - valid text-html accepted`() {
        assertDoesNotThrow { validateMimeType("text/html") }
    }

    @Test
    fun `mime type - valid application-json accepted`() {
        assertDoesNotThrow { validateMimeType("application/json") }
    }

    @Test
    fun `mime type - valid image-png accepted`() {
        assertDoesNotThrow { validateMimeType("image/png") }
    }

    @Test
    fun `mime type - rejects empty string`() {
        val ex = assertThrows<BridgeException> { validateMimeType("") }
        assertContains(ex.message.orEmpty(), "must not be empty")
        assertEquals("SCP-VALID-7011", ex.code)
    }

    @Test
    fun `mime type - rejects no slash`() {
        val ex = assertThrows<BridgeException> { validateMimeType("texthtml") }
        assertContains(ex.message.orEmpty(), "exactly one '/'")
    }

    @Test
    fun `mime type - rejects double slash`() {
        val ex = assertThrows<BridgeException> { validateMimeType("text/html/extra") }
        assertContains(ex.message.orEmpty(), "exactly one '/'")
    }

    @Test
    fun `mime type - rejects empty type part`() {
        val ex = assertThrows<BridgeException> { validateMimeType("/html") }
        assertContains(ex.message.orEmpty(), "both be non-empty")
    }

    @Test
    fun `mime type - rejects empty subtype part`() {
        val ex = assertThrows<BridgeException> { validateMimeType("text/") }
        assertContains(ex.message.orEmpty(), "both be non-empty")
    }

    @Test
    fun `mime type - rejects semicolon`() {
        val ex = assertThrows<BridgeException> { validateMimeType("text/html; charset=utf-8") }
        assertContains(ex.message.orEmpty(), "parameters")
    }

    @Test
    fun `mime type - rejects carriage return`() {
        val ex = assertThrows<BridgeException> { validateMimeType("text/html\r") }
        assertContains(ex.message.orEmpty(), "control character")
    }

    @Test
    fun `mime type - rejects line feed`() {
        val ex = assertThrows<BridgeException> { validateMimeType("text/html\n") }
        assertContains(ex.message.orEmpty(), "control character")
    }

    @Test
    fun `mime type - rejects null byte`() {
        val ex = assertThrows<BridgeException> { validateMimeType("text/\u0000html") }
        assertContains(ex.message.orEmpty(), "control character")
    }

    @Test
    fun `mime type - rejects C1 control character`() {
        val ex = assertThrows<BridgeException> { validateMimeType("text/\u0085html") }
        assertContains(ex.message.orEmpty(), "control character U+0085")
    }

    @Test
    fun `mime type - rejects non-tchar in type part`() {
        val ex = assertThrows<BridgeException> { validateMimeType("te xt/html") }
        assertContains(ex.message.orEmpty(), "type part contains invalid")
    }

    @Test
    fun `mime type - rejects non-tchar in subtype part`() {
        val ex = assertThrows<BridgeException> { validateMimeType("text/ht ml") }
        assertContains(ex.message.orEmpty(), "subtype part contains invalid")
    }

    @Test
    fun `mime type - accepts tchar special characters`() {
        assertDoesNotThrow { validateMimeType("application/vnd.foo+bar") }
    }

    // -- deploy_id ------------------------------------------------------------

    @Test
    fun `deploy id - valid simple accepted`() {
        assertDoesNotThrow { validateDeployId("deploy-1") }
    }

    @Test
    fun `deploy id - valid hex accepted`() {
        assertDoesNotThrow { validateDeployId("abc123def456") }
    }

    @Test
    fun `deploy id - valid underscore accepted`() {
        assertDoesNotThrow { validateDeployId("my_deploy_id") }
    }

    @Test
    fun `deploy id - valid mixed accepted`() {
        assertDoesNotThrow { validateDeployId("Deploy-2024_v1") }
    }

    @Test
    fun `deploy id - rejects empty string`() {
        val ex = assertThrows<BridgeException> { validateDeployId("") }
        assertContains(ex.message.orEmpty(), "must not be empty")
        assertEquals("SCP-VALID-7012", ex.code)
    }

    @Test
    fun `deploy id - rejects too long`() {
        val ex = assertThrows<BridgeException> { validateDeployId("a".repeat(129)) }
        assertContains(ex.message.orEmpty(), "exceeds 128 bytes")
    }

    @Test
    fun `deploy id - rejects spaces`() {
        val ex = assertThrows<BridgeException> { validateDeployId("deploy 1") }
        assertContains(ex.message.orEmpty(), "ASCII alphanumeric")
    }

    @Test
    fun `deploy id - rejects special characters`() {
        val ex = assertThrows<BridgeException> { validateDeployId("deploy@1") }
        assertContains(ex.message.orEmpty(), "ASCII alphanumeric")
    }

    @Test
    fun `deploy id - rejects slash`() {
        val ex = assertThrows<BridgeException> { validateDeployId("deploy/1") }
        assertContains(ex.message.orEmpty(), "ASCII alphanumeric")
    }
}
